packages/contracts/contracts/plugin/PluginInstaller.sol

```diff
@@ -0,0 +1,360 @@
+        // apply permissions on a dao..
+        DAO(payable(dao)).bulkOnMultiTarget(permissions);
+
+        // emit the event to connect plugin to the dao.
+        emit PluginInstalled(dao, plugin);
+
+        delete installPermissionHashes[hash];
+    }
+
+    // TODO: might we need to check when `deployUpdate` gets called, if plugin actually was installed ?
+    // Though problematic, since this check only happens when plugin updates from 1.0 to 1.x
+    // and checking it always would cost more... shall we still check it and how ?
+    function deployUpdate(
+        address dao,
+        PluginUpdateDeployInfo calldata updateInfo,
+        address[] calldata helpers, // helpers that were deployed when installing/updating the plugin.
+        bytes memory data // encoded per pluginManager's update ABI,
+    ) public returns (Permission.ItemMultiTarget[] memory, bytes memory) {
+        // check that plugin inherits from PluginUUPSUpgradeable
+        if (updateInfo.plugin.supportsInterface(type(PluginUUPSUpgradeable).interfaceId)) {
+            revert NotSupportsUpgradable();
+        }
+
+        // Implicitly confirms plugin managers are valid.
+        // ensure repo for plugin manager exists
+        if (!repoRegistry.entries(address(updateInfo.pluginManagerRepo))) {
+            revert PluginRepoNotExists();
+        }
+
+        (uint16[3] memory oldVersion, , ) = updateInfo.pluginManagerRepo.getVersionByPluginManager(
+            updateInfo.oldPluginManager
+        );
+
+        // Reverts if newPluginManager doesn't exist on the repo...
+        updateInfo.pluginManagerRepo.getVersionByPluginManager(updateInfo.newPluginManager);
+
+        // Check if helpers are correct...
+        // Implicitly checks if plugin was installed in the first place.
+        bytes32 oldHash = keccak256(
+            abi.encode(dao, updateInfo.oldPluginManager, updateInfo.plugin)
+        );
+
+        if (helperHashes[oldHash] != keccak256(abi.encode(helpers))) {
+            revert HelpersWrong();
+        }
+
+        delete helperHashes[oldHash];
+
+        // update deploy..
+        (
+            address[] memory activeHelpers,
+            bytes memory initData,
+            Permission.ItemMultiTarget[] memory permissions
+        ) = PluginManager(updateInfo.newPluginManager).update(
+                dao,
+                updateInfo.plugin,
+                helpers,
+                data,
+                oldVersion
+            );
+
+        // add new helpers for the future update checks
+        bytes32 newHash = keccak256(
+            abi.encode(dao, updateInfo.newPluginManager, updateInfo.plugin)
+        );
+        helperHashes[newHash] = keccak256(abi.encode(activeHelpers));
+
+        // check if permissions are corret.
+        updatePermissionHashes[newHash] = getPermissionsHash(permissions);
+
+        emit Updated(dao, activeHelpers, permissions);
+
+        return (permissions, initData);
+    }
+
+    function updatePlugin(
```

Comment:
The difference between deployUpdate and updatePlugin is not clear at all by the name

Reply:
Suggestion: `registerPluginUpdate` vs `applyPluginUpdate` ?